featurebyte/session/spark.py

```diff
@@ -4,72 +4,148 @@
 # pylint: disable=duplicate-code
 from __future__ import annotations
 
-from typing import Any, AsyncGenerator, Optional, OrderedDict
+from typing import Any, AsyncGenerator, Optional, OrderedDict, Union
+from typing_extensions import Annotated
 
+# pylint: disable=wrong-import-order
 import collections
+import os
+import subprocess
+import tempfile
 
 import pandas as pd
 import pyarrow as pa
 from pandas.core.dtypes.common import is_datetime64_dtype, is_float_dtype
 from pyarrow import Schema
-from pydantic import Field
+from pydantic import Field, PrivateAttr
 from pyhive.exc import OperationalError
 from pyhive.hive import Cursor
+from thrift.transport.TTransport import TTransportException
 
-from featurebyte.enum import DBVarType, SourceType
+from featurebyte.enum import DBVarType, SourceType, StorageType
 from featurebyte.logging import get_logger
+from featurebyte.models.credential import AccessTokenCredential, KerberosKeytabCredential
 from featurebyte.session.base_spark import BaseSparkSession
 from featurebyte.session.hive import AuthType, HiveConnection
+from featurebyte.session.simple_storage import WebHDFSStorage
 
 logger = get_logger(__name__)
 
 
+SparkDatabaseCredential = Annotated[
+    Union[KerberosKeytabCredential, AccessTokenCredential],
+    Field(discriminator="type"),
+]
+
+
 class SparkSession(BaseSparkSession):
     """
     Spark session class
     """
 
     _no_schema_error = OperationalError
+    _connection: Optional[HiveConnection] = PrivateAttr(None)
 
     port: int
     use_http_transport: bool
     use_ssl: bool
-    access_token: Optional[str]
     source_type: SourceType = Field(SourceType.SPARK, const=True)
+    database_credential: Optional[SparkDatabaseCredential]
 
     def __init__(self, **data: Any) -> None:
-        super().__init__(**data)
-
         auth = None
         scheme = None
+        access_token = None
+        kerberos_service_name = None
+
+        super().__init__(**data)
+
+        if self.database_credential:
```

Comment:
would it be cleaner to have separate `SparkSession#fromToken` and `SparkSession#fromKerberos` methods, instead of trying to overload the constructor this way?

Reply:
It makes sense if we are instantiating these classes manually. However since these classes are instantiated by the session manager, it will make the logic there more complex.